akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergePrioritizedSpec.scala

```diff
@@ -30,120 +30,122 @@ class GraphMergePrioritizedSpec extends TwoStreamsSetup {
+        threes shouldEqual 0
+        (ones / twos).round shouldBe 2
       }
-
-      val ones = collected.count(_ == 1).toDouble
-      val twos = collected.count(_ == 2).toDouble
-      val threes = collected.count(_ == 3)
-
-      threes shouldEqual 0
-      (ones / twos).round shouldBe 2
     }
   }
 
-  private def threeSourceMerge[T](source1: Source[T, NotUsed], source2: Source[T, NotUsed], source3: Source[T, NotUsed], priorities: Seq[Int], probe: ManualProbe[T]) = {
-    RunnableGraph.fromGraph(GraphDSL.create(source1, source2, source3)((_, _, _)) { implicit b ⇒ (s1, s2, s3) ⇒
+  private def threeSourceMergeSource[T](source1: Source[T, NotUsed], source2: Source[T, NotUsed], source3: Source[T, NotUsed], priorities: Seq[Int]) = {
+    Source.fromGraph(GraphDSL.create(source1, source2, source3)((_, _, _)) { implicit b ⇒ (s1, s2, s3) ⇒
       val merge = b.add(MergePrioritized[T](priorities))
-      // introduce a delay on the consuming side making it more likely that
-      // the actual prioritization happens and elements does not just pass through
-      val delayFirst = b.add(Flow[T].initialDelay(50.millis))
       s1.out ~> merge.in(0)
       s2.out ~> merge.in(1)
       s3.out ~> merge.in(2)
-      merge.out ~> delayFirst ~> Sink.fromSubscriber(probe)
-      ClosedShape
+      SourceShape(merge.out)
     })
   }
 }
+
+object Slice {
+  final val Success = new Error("Success!")
+  final class AssertionFailed(neededMore: Int, target: Int, maxTries: Int, val exceptions: Seq[Exception]) extends Exception(
+    s"Was not able to satisfy the assertion at least [$target] times in [$maxTries] tries. [$neededMore] more successful assertions needed."
+  )
+
+  def assert[T](source: Source[T, _], duration: FiniteDuration, maxTries: Int, target: Int)(assertion: Seq[T] ⇒ Any)(implicit mat: Materializer): Unit = {
+    val timeout = duration * (maxTries + 2)
+
+    val result = source
+      // introduce an async boundary on the consuming side making it more likely that
+      // the actual prioritization happens and elements does not just pass through
+      .async
+      .groupedWithin(Int.MaxValue, duration)
+      .runFold((maxTries, target, Seq.empty[Exception])) {
+        // no more tries left
+        case ((0, more, exceptions), _) ⇒ throw new AssertionFailed(more, target, maxTries, exceptions)
+
+        // reached the target successful assertions
+        case ((_, 0, _), _)             ⇒ throw Success
+
+        case ((max, success, exceptions), collected) ⇒ try {
+          assertion(collected)
+          (max - 1, success - 1, exceptions)
+        } catch {
+          case ex: Exception ⇒
+            (max - 1, success, ex +: exceptions)
+        }
+      }
+
+    Await.result(result.failed, timeout) match {
+      case ex if ex.getCause != null && ex.getCause == Success ⇒
+      case ex: AssertionFailed ⇒
+        ex.exceptions.foreach(_.printStackTrace)
+        throw ex
+      case ex ⇒ throw ex
+    }
+  }
+}
```

Comment:
wouldn't something like `retryAtMost(n) { testblock }`, completely unaware of streams, be a simpler way to do the same?

Reply:
similar to scalatests' Eventually? That does sound appealing...

Reply:
or `awaitAssert`